Fix deprecated rename issue

Kyle-Ye opened the pull request:
When I upgrade from v2.1.4 to v2.4.3, some API got changed.

And I found the renamed suggestion seems to be wrongly placed.

The original "HTMLTemplate" needs a context which corresponds to current "View".
The original "HTMLPage" does not need a context which corresponds to current "Page"

mattesmohr commented:
I disagree on that. Truly I have could done the renaming a bit better to prevent confusion. Sorry for the inconvenience. 

Besides the renaming I have moved the context property from HTMLTemplate (now Page) to HTMLPage (now View), cause I see the Page more as a wrapper, maybe comparable to `WindowGroup`, I guess. 

And I agree, that sometimes you want to have a context on `Page` too, something comparable to `EnvironmentObject`. I want to implement it in the future. 

I wrote down my idea in the instructions: [Layouts.md](https://github.com/vapor-community/HTMLKit/blob/main/Instructions/Essential/Layouts.md) and I also I mentioned to work on that in the discussion: [2.5.0](https://github.com/vapor-community/HTMLKit/discussions/56).

Sorry I cant approve your PR, since the deprecated annotations are right. But I appreciate your help and concerns.

Is there anything I can help you with?

Kyle-Ye commented:
Got it, but for the users upgraded from v2.1.4 to v2.4.3.
Follow the recommended renaming will not fix the compiler error.
(Renaming "HTMLPage" to "View" and "HTMLTemplate" to "Page")

Instead if I just rename "HTMLPage" to "Page" and "HTMLTemplate" to "View", it fits rather well and without some major change.

<img width="1532" alt="截屏2022-01-16 21 49 53" src="https://user-images.githubusercontent.com/43724855/149662854-cd002ab0-039f-41f8-a62a-526701ae177f.png">

<img width="1236" alt="截屏2022-01-16 21 49 33" src="https://user-images.githubusercontent.com/43724855/149662857-7e4b86c0-b587-40f9-8d08-16a50a7f3b2b.png">

mattesmohr commented:
Can you tell me the Error, you are hitting? I try to reproduce the issue.

mattesmohr commented:
I think I got you. There is `render(with: )` method missing, right?

Kyle-Ye commented:
> Can you tell me the exact error, you are hitting? I will try to reproduce the issue.

The original "HTMLTemplate" needs a context which corresponds to current "View".
The original "HTMLPage" does not need a context which corresponds to current "Page"

I think you might have reversed the renaming suggestion for "View" and "Page".

See #45 and #46

Kyle-Ye commented:
> Besides the renaming I have moved the context property from HTMLTemplate (now Page) to HTMLPage (now View)

Where is the change? I checked the relevant git history and couldn't find it. If you can point it for me, I'd be very appreciated. ❤️

mattesmohr commented:
> > Can you tell me the exact error, you are hitting? I will try to reproduce the issue.
> 
> The original "HTMLTemplate" needs a context which corresponds to current "View". The original "HTMLPage" does not need a context which corresponds to current "Page"
> 
> I think you might have reversed the renaming suggestion for "View" and "Page".
> 
> See #45 and #46

No, that's fine. 

At the moment the vapor extension (HTMLKitVaporProvider) offers no render-method with a context if its a `Page`. I guess that's the issue you are encounter in Xcode. 

`AppPrivacyTemplate` should be your wrapper and `HomePage` should be your view, right?

But as you do right now (the turnaround), it is the only workaround at the moment. So Its okay if you use View on AppPrivacyTemplate and I will consider it in the release 2.5.0.

mattesmohr commented:
> > Besides the renaming I have moved the context property from HTMLTemplate (now Page) to HTMLPage (now View)
> 
> Where is the change? I checked the relevant git history and couldn't find it. If you can point it for me, I'd be very appreciated. ❤️

Its in this commit [33294017afc0567392f8f0e10945d984015e0c37](https://github.com/vapor-community/HTMLKit/pull/46/commits/33294017afc0567392f8f0e10945d984015e0c37)

Kyle-Ye commented:
> > Besides the renaming I have moved the context property from HTMLTemplate (now Page) to HTMLPage (now View)

> Where is the change? I checked the relevant git history and couldn't find it. If you can point it for me, I'd be very appreciated. 

I think we might be overcomplicating things

It is easy. See the diff on Formulas.swift between 45 and 46.

Isn't it just rename "HTMLTemplate" to "View" and "HTMLPage" to "Page"?

#45 
https://github.com/vapor-community/HTMLKit/blob/81091445a6fcca08616c060c0e1089272640fed4/Sources/HTMLKit/Internal/Formulas.swift#L4-L47

#46 
https://github.com/vapor-community/HTMLKit/blob/7743c17015bd6e76982c273e48f5e05d23a459f7/Sources/HTMLKit/External/Formulas.swift#L4-L47


I got confused by what you say🤔️.

And FYI this is my usage. (Although I think this is irrelevant to what I'm talking about)
```swift
try app.htmlkit.add(view: AppPrivacyTemplate())
app.get { req in
    HomePage().render(for: req)
}

app.get("AppA") { req in
    AppPrivacyTemplate().render(with: .appA, for: req)
}

app.get("AppB") { req in
    AppPrivacyTemplate().render(with: .appB, for: req)
}
```

Kyle-Ye commented:
> > > Besides the renaming I have moved the context property from HTMLTemplate (now Page) to HTMLPage (now View)
> > 
> > 
> > Where is the change? I checked the relevant git history and couldn't find it. If you can point it for me, I'd be very appreciated. ❤️
> 
> Its in this commit [33294017afc0567392f8f0e10945d984015e0c37](https://github.com/vapor-community/HTMLKit/pull/46/commits/33294017afc0567392f8f0e10945d984015e0c37) at line 32

Emm, I still can not get it. The whole MR is meant to change "HTMLPage" to "Page".

It doesn't matter if you only change the relative position.

<img width="1231" alt="截屏2022-01-16 22 40 33" src="https://user-images.githubusercontent.com/43724855/149664795-93408d05-a31f-4408-bef5-1886e46da894.png">

Also here, why not we just change L55 and L62 of "HTMLTemplate" to "View" and "HTMLPage" to "Page". If so, we only change 2 lines. And here we changed so many lines and is just the same result.(For me it seems like it just changed the relate position of "View" and "Page")

Kyle-Ye commented:
Checked the full conversation of #46 
Though it's written as
```
The HTMLTemplate will become Page.
The HTMLPage will become View.
```
But for the end user's API usage, it looks like "Page" was more like "HTMLPage" and "View" was more like "HTMLTemplate".

mattesmohr commented:
> Isn't it just rename "HTMLTemplate" to "View" and "HTMLPage" to "Page"?

Nope, it is not. Cause I tried to explain my intention/idea of `Page` and `View` in my previous comment.

```
Page {
      View {
         ViewContext
      }
}
```

In your case `AppPrivacyTemplate` should be `Page`. And `HomePage` should be `View`. And `View` should hold the `Context`. And that's how the deprecation annotation tells you to do.

But you are right, you are hitting issues, when you do as the annotation says, cause I forgot to consider the possibility, that's the user want to add a `PageContext`. 

```
Page {
      PageContext
      View {
      }
}
```

```swift

/// that's is not working if AppPrivacyTemplate is `Page`. Cause the .render(with) method exists only 
/// if the object is type of `View`.

app.get("AppA") { req in
    AppPrivacyTemplate().render(with: .appA, for: req)
}
```

So I have to extend the vapor provider. But it will take a while. So keep your code as it is (as a workaround). 

Sorry, I don't know how to explain it differently ;-).

Kyle-Ye commented:
Got it, thank you for your detailed explanation.
